lazy_dataset/core.py

```diff
@@ -1640,26 +1757,69 @@ def indexable(self):
     def __iter__(self):
         buckets = list()
         for i, sample in enumerate(self.input_dataset):
-            value = min(self.key(sample), self.max_value)
+            value = self.key(sample)
             found_bucket = False
-            for i, (bucket, min_value, max_value) in enumerate(buckets):
-                if min_value <= value <= max_value:
+            expired = set()
+            for j, (
+                bucket, creation_idx, lower_bound, upper_bound, max_value
+            ) in enumerate(buckets):
+                if (
+                    self.expiration is not None
+                    and (i - creation_idx) / self.batch_size > self.expiration
+                ):
+                    expired.add(j)
+                    if not self.drop_incomplete:
+                        if self.sort_by_key:
+                            bucket = sorted(bucket, key=self.key, reverse=True)
+                        yield bucket
+                    else:
+                        # ToDo: maybe add warning
+                        pass
+                elif not found_bucket and lower_bound <= value <= upper_bound:
                     bucket.append(sample)
-                    if len(bucket) >= self.batch_size:
-                        buckets.pop(i)
+                    max_value = max(max_value, value)
+                    if (
+                        len(bucket) >= self.batch_size
+                        or (
+                            self.max_total_size is not None
+                            and (len(bucket) * max_value) > self.max_total_size
+                        )
+                    ):
+                        expired.add(j)
+                        if self.sort_by_key:
+                            bucket = sorted(bucket, key=self.key, reverse=True)
                         yield bucket
```

Comment:
Is here a break missing?

Reply:
I keep iterating to yield buckets which have exceeded total_size_threshold

Reply:
Ah, my fault, I overread the `elif not found_bucket`